python/ray/experimental/ddp/ddp.py

```diff
@@ -0,0 +1,518 @@
+                [
+                    actor.get_grad_to_reduce.bind(grads[i])
+                    for i, actor in enumerate(actors)
+                ]
+            )
+            updates = [
+                actor.update.bind(j, reduced_grad)
+                for actor, reduced_grad in zip(actors, reduced_grads)
+            ]
+            output.append(updates)
+        ends = [
+            actor.finish_train.bind(
+                *[output[j][i] for j in reversed(range(num_layers))]
+            )
+            for i, actor in enumerate(actors)
+        ]
+        dag = MultiOutputNode(ends)
+
+    compiled_dag = dag.experimental_compile()
+    it = config.it
+    adag_ddp_weights = []
+    elapses = []
+    for i in range(it):
+        x, y = generate_input_output(config)
+        xs = torch.tensor_split(x, num_actors)
+        ys = torch.tensor_split(y, num_actors)
+        x = time.perf_counter()
+        ref = compiled_dag.execute(*xs, *ys)
+        cur_iter_weights = ray.get(ref)
+        end = time.perf_counter()
+        adag_ddp_weights.append(cur_iter_weights)
+        elapse = end - x
+        elapses.append(elapse)
+
+    print_elapses(elapses, "adag ddp")
+    compiled_dag.teardown()
+
+    for actor in actors:
+        ray.kill(actor)
+
+    return adag_ddp_weights
+
+
+def run_torch(config):
+    """Run pytorch without DDP and return the weights."""
+    device = "cuda:0"
+    model = LayeredModel(
+        config.layer_size, config.num_layers, device, config.dtype, config.lr
+    )
+    criterion = model.criterion
+    optimizer = optim.SGD(model.parameters(), lr=model.lr)
+
+    torch_weights = []
+    elapses = []
+    for i in range(config.it):
+        x, y = generate_input_output(config)
+        x = x.to(device)
+        y = y.to(device)
+        start = time.perf_counter()
+        optimizer.zero_grad()
+        pred: torch.Tensor = model(x)
+        loss: torch.Tensor = criterion(pred, y)
+        loss.backward()
+        optimizer.step()
+        end = time.perf_counter()
+
+        cur_iter_weights = []
+        for i in range(0, len(model.layers), 2):
+            layer: torch.nn.Linear = model.layers[i]
+            cur_iter_weights.append(torch.clone(layer.weight))
+        torch_weights.append(cur_iter_weights)
+
+        elapse = end - start
+        elapses.append(elapse)
+
+    print_elapses(elapses, "torch")
+
+    return torch_weights
+
+
+def run_torch_ddp(torch_weights, config: Config):
+    """Run DDP with `torch.distributed`. Check and compare the weights."""
+    n_gpus = torch.cuda.device_count()
+    assert (
+        n_gpus >= config.num_actors
+    ), f"Requires at least {config.num_actors} GPUs to run, but got {n_gpus}"
+    world_size = n_gpus
+    mp.spawn(
+        torch_dist_run,
+        args=(world_size, torch_weights, config),
+        nprocs=world_size,
+        join=True,
+    )
+
+
+def torch_dist_run(rank, world_size, torch_weights, config: Config):
```

Comment:
What's the difference of `run_torch_ddp` and `torch_dist_run`? The comments look similar.

Reply:
Renamed `torch_dist_run` to `torch_ddp_per_process`.